### PR TITLE
Bugfix/footstep plan action

### PR DESCRIPTION
**Pull request by calvertdw**

### CODE REVIEW

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java [hunk 1]
```diff
@@ -91,6 +93,20 @@ else if (definition.getPlannerInitialStanceSide().getValue() == InitialStanceSid
...
+      {
+         // At beginning, first turn in place to face the direction that the goal stance faces
+         startMidFeetPose.interpolate(startFootPoses.get(RobotSide.LEFT), startFootPoses.get(RobotSide.RIGHT), 0.5);
+         goalMidFeetPose.interpolate(goalFootPoses.get(RobotSide.LEFT), goalFootPoses.get(RobotSide.RIGHT), 0.5);
+         startTurnedToMatchGoalFacing.set(startMidFeetPose);
+         startTurnedToMatchGoalFacing.getOrientation().set(goalMidFeetPose.getOrientation());
+         footstepPlannerRequest.getBodyPathWaypoints().add(startMidFeetPose);
+         footstepPlannerRequest.getBodyPathWaypoints().add(startTurnedToMatchGoalFacing);
+         footstepPlannerRequest.getBodyPathWaypoints().add(goalMidFeetPose);
+      }
```
Comment (calvertdw) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java:120]: I added this block which makes sure when planning backwards, the robot doesn't try to turn around just to take a few steps back.

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+import us.ihmc.footstepPlanning.FootstepPlannerRequest;
+import us.ihmc.footstepPlanning.FootstepPlanningModule;
+import us.ihmc.footstepPlanning.graphSearch.graph.visualization.BipedalFootstepPlannerNodeRejectionReason;
+import us.ihmc.footstepPlanning.graphSearch.parameters.InitialStanceSide;
+import us.ihmc.footstepPlanning.log.FootstepPlannerLogger;
+import us.ihmc.footstepPlanning.tools.FootstepPlannerRejectionReasonReport;
+import us.ihmc.robotics.robotSide.RobotSide;
+import us.ihmc.robotics.robotSide.SideDependentList;
+
+public class FootstepPlanActionPlanningThread
```
Comment (ds58) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java:19]: I dont see a thread, why is it labeled as a thread?

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+   {
+      ++started;
+
+      for (RobotSide side : RobotSide.values)
+      {
+         startFootPoses.get(side).setFromReferenceFrame(syncedRobot.getReferenceFrames().getSoleFrame(side));
+         goalFootPoses.get(side).set(liveGoalFeetPoses.get(side));
+      }
+
+      Thread thread = new Thread(() -> plan(started), getClass().getSimpleName() + started);
```
Comment (calvertdw) [inline ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java]: TODO: Need to add exception handling

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java (hunk 2 above)
Comment (ds58) [reply ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java:19]: Nevermind, I see you're creating a thread each update loop. Can you change it to where there is only one thread, as the class name implies, and have the `triggerPlan` method submit new jobs to run?

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java (hunk 1 above)
Comment (LuigiPenco93) [reply ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java:120]: Isn't this handled by the UseTurnWalkTurn and A star search?

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java (hunk 2 above)
Comment (LuigiPenco93) [reply ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java:19]: I'm also curious to know if there's a reason to create several plans (threads) for one FootstepPlanActionPlanningThread

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java (hunk 1 above)
Comment (calvertdw) [reply ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java:120]: The use turn walk turn checkbox literally is the setting for whether to use A*:
```
footstepPlannerRequest.setPerformAStarSearch(!definition.getPlannerUseTurnWalkTurn().getValue());
```

The turn walk turn planner however take way more unnecessary steps than A*, which is why it should be preferred. I'm gonna makes some changes to clear this up.

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java (hunk 3 above)
Comment (calvertdw) [reply ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java]: Done

File: ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java (hunk 2 above)
Comment (calvertdw) [reply ihmc-high-level-behaviors/src/main/java/us/ihmc/behaviors/sequence/actions/FootstepPlanActionPlanningThread.java:19]: Yeah so preview plans are triggered once a second while the action is next for execution. Sometimes a prior plan can run long and a newer one can be done faster, so we want to allow to plan in parallel to improve the user experience.

Also the plan for execution is done always once when the action execution is triggered. We want this to be very independent of the preview planning.

**Review by LuigiPenco93 — APPROVED** [review]

**Pull request merged**